app/src/main/kotlin/compiler/analysis/InitializationVerifier.kt

```diff
@@ -0,0 +1,155 @@
+        nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
+        defaultParameterMapping: Map<Ref<Function.Parameter>, Variable>,
+        diagnostics: Diagnostics
+    ) {
+        fun verifyInitialization(
+            node: AstNode,
+            initializedVariables: MutableSet<Ref<NamedNode>>,
+            traversedFunctions: MutableSet<Ref<Function>> // so that we don't analyze calls of the same function multiple times
+        ) {
+            when (node) {
+                is Statement.Evaluation -> verifyInitialization(node.expression, initializedVariables, traversedFunctions)
+                is Statement.VariableDefinition -> verifyInitialization(node.variable, initializedVariables, traversedFunctions)
+                is Global.VariableDefinition -> verifyInitialization(node.variable, initializedVariables, traversedFunctions)
+                is Statement.FunctionDefinition -> {
+                    verifyInitialization(node.function, initializedVariables, traversedFunctions)
+                }
+                is Global.FunctionDefinition -> verifyInitialization(node.function, initializedVariables, traversedFunctions)
+                is Statement.Assignment -> {
+                    verifyInitialization(node.value, initializedVariables, traversedFunctions)
+                    // the variable will be initialized after assignment, not before
+                    initializedVariables.add(nameResolution[Ref(node)]!!)
+                }
+                is Statement.Block -> {
+                    node.block.forEach { verifyInitialization(it, initializedVariables, traversedFunctions) }
+                }
+                is Statement.Conditional -> {
+                    verifyInitialization(node.condition, initializedVariables, traversedFunctions)
+                    val initializedVariablesWhenTrue = initializedVariables.toMutableSet()
+                    val traversedFunctionsWhenTrue = traversedFunctions.toMutableSet()
+                    node.actionWhenTrue.forEach {
+                        verifyInitialization(
+                            it, initializedVariablesWhenTrue,
+                            traversedFunctionsWhenTrue
+                        )
+                    }
+
+                    val initializedVariablesWhenFalse = initializedVariables.toMutableSet()
+                    val traversedFunctionsWhenFalse = traversedFunctions.toMutableSet()
+                    node.actionWhenFalse?.forEach {
+                        verifyInitialization(
+                            it, initializedVariablesWhenFalse,
+                            traversedFunctionsWhenFalse
+                        )
+                    }
+
+                    val initializedVariablesWhenBoth = initializedVariablesWhenTrue intersect initializedVariablesWhenFalse
+                    val traversedFunctionsWhenBoth = traversedFunctionsWhenTrue intersect traversedFunctionsWhenFalse
+                    initializedVariables.clear()
```

Comment:
If I'm not mistaken, these `clear`s are not needed - if so, why don't we remove them?